reconstruction/ecoli/dataclasses/process/metabolism_new_pathway.py

```diff
@@ -0,0 +1,328 @@
+                    # Molecules that are required to produce ATP with the independent stoich matrix
+                    self.atp_reaction_reactant_mask = self.dependency_matrix[:, self.independent_molecules_atp_index] < 0
+
+            else:
+                print('There is no metabolic reaction to be added')
+        except Exception as e:
+            print('No extra genes were added')
+            print(repr(e))
+            print(e)
+
+
+    def __getstate__(self):
+        """Return the state to pickle, omitting derived attributes that
+        __setstate__() will recompute, esp. the ode_derivatives
+        that don't pickle.
+        """
+        try:
+            return data.dissoc_strict(self.__dict__, (
+                'symbolic_rates','symbolic_rates_jacobian',
+			'_rates', '_rates_jacobian',
+			'_stoich_matrix'))
+
+        except:
+            return 0
+
+
+    def __setstate__(self, state):
+        """Restore instance attributes, recomputing some of them."""
+        try:
+            self.__dict__.update(state)
+            self._populate_derivative_and_jacobian()
+        except:
+            return 0
+
+    def stoich_matrix(self):
+        '''
+        Builds stoichiometry matrix
+        Rows: molecules
+        Columns: reactions
+        Values: reaction stoichiometry
+        '''
+        shape = (self._stoichMatrixI.max() + 1, self._stoichMatrixJ.max() + 1)
+        out = np.zeros(shape, np.float64)
+        out[self._stoichMatrixI, self._stoichMatrixJ] = self._stoichMatrixV
+        return out
+
+    def mass_matrix(self):
+        '''
+        Builds stoichiometry mass matrix
+        Rows: molecules
+        Columns: reactions
+        Values: molecular mass
+        '''
+        shape = (self._stoichMatrixI.max() + 1, self._stoichMatrixJ.max() + 1)
+        out = np.zeros(shape, np.float64)
+        out[self._stoichMatrixI, self._stoichMatrixJ] = self._stoich_matrix_mass
+        return out
+
+    def mass_balance(self):
+        '''
+        Sum along the columns of the massBalance matrix to check for reaction mass balance
+        '''
+        return np.sum(self.balance_matrix, axis=0)
+
+    def _populate_derivative_and_jacobian(self):
+        '''Compile callable functions for computing the derivative.'''
+        self._make_derivative()
+
+        self._rates = build_ode.derivatives(self.symbolic_rates)
+        self._rates_jacobian = build_ode.derivatives_jacobian(self.symbolic_rates_jacobian)
+        self._stoich_matrix = self.stoich_matrix()  # Matrix is small and can be cached for derivatives
+
+
+    def _make_y_dy(self):
+        '''This is building the right hand side of the ODE system using mass action kinetics.'''
+
+        S = self.stoich_matrix()
+
+        yStrings = ["y[%d]" % x for x in range(S.shape[0])]
+
+        y = sp.symbols(yStrings)
+
+        rates = []
+        for colIdx in range(S.shape[1]):
+            #here we assume that the substrate is the first element in the reaction
```

Comment:
This is a pretty strong assumption that I think needs to be clearly highlighted elsewhere (maybe in the new reactions flat file as well). Ideally we should be able to clearly specify which metabolite is the substrate that goes into the MM equation.